File: test/end-to-end/VectorClass/Python/vector.py
#
from math import floor, ceil
class Vector:

    def __init__(self):
        self.__capacity = 0
        self.__data = []
        self.__length = 0

    def at(self, index):
        if index >= self.__length:
            raise Exception("Index out of bounds: {0} is greater than {1}.".format(index, self.__length))

        return self.__data[index]

    def get_capacity(self):
        return self.__capacity

    def get_length(self):
        return self.__length

    def get_first(self):
        if self.__capacity == 0:
            raise Exception("Cannot get first from empty vector.")

        return self.__data[0]

    def get_last(self):
        if self.__capacity == 0:
            raise Exception("Cannot get last from empty vector.")

        return self.__data[self.__length - 1]

    def ensure_capacity(self, capacity):
        if capacity <= self.__capacity:
            return

        newCapacity = floor(ceil((capacity / 2.0)) * 3)
        oldData = self.__data
        self.__capacity = newCapacity
        self.__data = [None] * newCapacity

        for i in range(0, len(oldData)):
            self.__data[i] = oldData[i]

    def push(self, item):
        self.ensure_capacity(self.__length + 1)

        self.__data[self.__length] = item
        self.__length += 1

    def resize(self, length):
        if length <= self.__length:
            self.__length = length
            return

        self.ensure_capacity(length)
        self.__length = length

    def to_array(self):
        array = [None] * self.__length

        for i in range(0, self.__length):
            array[i] = self.__data[i]

        return array

    def to_list(self):
        list = [None] * self.__length

        for i in range(0, self.__length):
            list[i] = self.__data[i]

        return list

    def to_set(self):
        individuals = set()

        for i in range(0, self.__length):
            individuals.add(self.__data[i])

        return individuals
#

File: test/end-to-end/VectorClass/Java/Vector.java
package vectorclass;

import java.util.Arrays;
import java.util.HashMap;
import java.util.HashSet;
import java.util.ArrayList;
import java.util.Comparator;

public class Vector<T> {
    private T[] data;
    private Integer capacity;
    private Integer length;

    public Vector() {
        this.capacity = 0;
        this.data = (T[]) new Object[0];
        this.length = 0;
    }

    public T at(Integer index) throws Exception {
        if (index >= this.length) {
            throw new Exception(String.format("Index out of bounds: %d is greater than %d.", index, this.length));
        }

        return this.data[index];
    }

    public Integer getCapacity() {
        return this.capacity;
    }

    public Integer getLength() {
        return this.length;
    }

    public T getFirst() throws Exception {
        if (this.capacity == 0) {
            throw new Exception("Cannot get first from empty vector.");
        }

        return this.data[0];
    }

    public T getLast() throws Exception {
        if (this.capacity == 0) {
            throw new Exception("Cannot get last from empty vector.");
        }

        return this.data[this.length - 1];
    }

    public void ensureCapacity(Integer capacity) {
        if (capacity <= this.capacity) {
            return;
        }

        Integer newCapacity = (int)Math.floor(Math.ceil((capacity / 2.0)) * 3);
        T[] oldData = this.data;
        this.capacity = newCapacity;
        this.data = (T[]) new Object[newCapacity];

        for (Integer i = 0; i < oldData.length; i += 1) {
            this.data[i] = oldData[i];
        }
    }

    public void push(T item) {
        this.ensureCapacity(this.length + 1);

        this.data[this.length] = item;
        this.length += 1;
    }

    public void resize(Integer length) {
        if (length <= this.length) {
            this.length = length;
            return;
        }

        this.ensureCapacity(length);
        this.length = length;
    }

    public T[] toArray() {
        T[] array = (T[]) new Object[this.length];

        for (Integer i = 0; i < this.length; i += 1) {
            array[i] = this.data[i];
        }

        return array;
    }

    public ArrayList<T> toList() {
        ArrayList<T> list = new ArrayList<T>(this.length);

        for (Integer i = 0; i < this.length; i += 1) {
            list.set(i, this.data[i]);
        }

        return list;
    }

    public HashSet<T> toSet() {
        HashSet<T> individuals = new HashSet<T>();

        for (Integer i = 0; i < this.length; i += 1) {
            individuals.add(this.data[i]);
        }

        return individuals;
    }
}
